onFlush and postFlush Entity Listeners

alexgurrola opened the pull request:
Simply check if an onFlush or postFlush entity listener is subscribed
for the current class in the entity manager during dispatch.

This is a followup to #6345, #5822, and #5582.

theofidry commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -3294,18 +3294,48 @@ private function performCallbackOnCachedPersister(callable $callback)
+                                $subscribedSystems[$entityClass] = $this->listenersInvoker->getSubscribedSystems($class, $event) & ~ListenersInvoker::INVOKE_MANAGER;
+                            }
+                            $invoke = $subscribedSystems[$entityClass];
+                            if (ListenersInvoker::INVOKE_NONE !== $invoke) {
```
doing early returns here would avoid to go to 6 levels of nesting

alexgurrola replied:
Wouldn't a return close the function and not allow it to continue to the next Entity Listener?

theofidry replied:
`continue`* (same idea)

alexgurrola replied:
I added all the continues rather than the nested if statements.

theofidry commented:
It missing some tests as well and it would be nice to check the performance impact as well.

lcobucci commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -3294,18 +3294,52 @@ private function performCallbackOnCachedPersister(callable $callback)
+                    continue;
+                }
+                $hash = (string) $hash;
+                if (!isset($this->identityMap[$className][$hash])) {
```
CS not followed https://github.com/doctrine/doctrine2/blob/master/CONTRIBUTING.md#coding-standard

alexgurrola replied:
@lcobucci Good eye.  I fixed the spacing there.

lcobucci commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -3294,18 +3294,52 @@ private function performCallbackOnCachedPersister(callable $callback)
+                }
+                $entityClass = get_class($entity);
+                $class = $this->em->getClassMetadata($entityClass);
+                if (!isset($subscribedSystems[$entityClass])) {
```
CS not followed https://github.com/doctrine/doctrine2/blob/master/CONTRIBUTING.md#coding-standard

alexgurrola replied:
@lcobucci I fixed the spacing here as well.

lcobucci reviewed (changes requested):
Please extract some methods to make the code more readable, the `UnitOfWork` is already a mess...

Obviously we need tests we also need to check the performance impact (@guilhermeblanco and @Ocramius are the best ones to say something regarding that).

And finally, as I asked before, please **stop creating new PRs for the same thing**, just rebase your branch. What your doing doesn't help the review, in fact it makes it more difficult since we have to check more than one place to track all the review comments.

ostrolucky reviewed (changes requested):
Stopped the further review because as I read further, I now doubt this actually works. Needs tests definitely.

alexgurrola commented:
@lcobucci When I tried to rebase the last Pull Request, it wouldn't let me push to the repo, and the other one that did messed up all the commits in the log, so I had to delete the branch to recreate it, which ended up closing the Pull Request.

alexgurrola commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -3294,18 +3294,52 @@ private function performCallbackOnCachedPersister(callable $callback)
+                if (!$hash) {
+                    continue;
+                }
+                $hash = (string) $hash;
```
I was having problems with the tests trying to pass in Objects as IDs, so I implemented the Type Cast to pass said tests.

alexgurrola commented:
@ostrolucky This pull request is the only reason we maintain a fork at https://github.com/Sitetheory/doctrine2 for composer to get these features on our web servers, which do require such functionality to correctly time various sets of business logic, and have worked perfectly fine with our fork for over a year.  That said, I'm certain there are better ways of providing this level of functionality, but as @lcobucci stated, the `UnitOfWork.php` is already a mess, so this sort of implementation has become quite difficult for us to build, maintain, and optimize without said expertise in this particular file.

alexgurrola commented on lib/Doctrine/ORM/UnitOfWork.php:
```diff
@@ -3294,18 +3294,52 @@ private function performCallbackOnCachedPersister(callable $callback)
+        $subscribedSystems = [];
+        foreach ($this->entityChangeSets as $oid => $entityChangeSet) {
+            foreach ($this->identityMap as $className => $entities) {
+                $hash = (isset($this->entityIdentifiers[$oid], $this->entityIdentifiers[$oid]['id'])) ? $this->entityIdentifiers[$oid]['id'] : null;
```
I may have some time later today or tomorrow to give you a better answer than this one, but the main reason that I was using the nested `foreach` loop was due to an attempt at creating logic similar to that of the other Entity Listeners, but on the `onFlush` and `postFlush` events, which do not digest each entity as easily as the other events, which can piggyback off the loops already available.

alexgurrola replied:
@ostrolucky I've addressed all of your issues, and it is now much more concise.  👍

lcobucci commented:
> @lcobucci When I tried to rebase the last Pull Request, it wouldn't let me push to the repo, and the other one that did messed up all the commits in the log, so I had to delete the branch to recreate it, which ended up closing the Pull Request.

@alexgurrola I'm sorry if what I said was a bit harsh, I didn't mean to be rude.

Since you didn't say the reason why you open/closed PRs for the same thing I understood it was your way to do (and it was really confusing for me). So if you need help regarding rebase let us know, we can help you on that 😉.

alexgurrola commented:
I have fixed any CS issues and cut down on the resources utilized by this routine, drastically.  It should now be much more efficient, solid, and easy to maintain in the future.  If you need anything else, let me know.  😄 

Reviewers: @lcobucci @theofidry @ostrolucky @Ocramius

alexgurrola commented:
The only check that was unsuccessful seems to be a Travis timeout?

Ocramius commented:
@alexgurrola restarted that failed task

alexgurrola commented:
@lcobucci Your requested changes are holding the merge back.  Can you review the code and see if it addresses all complaints?

Thanks,
-- Alex

Ocramius reviewed (changes requested):
@alexgurrola tests weren't added: tests describing the change and verifying this new behavior are required

mwansinck commented:
Hi @alexgurrola, any progress on this?

alexgurrola committed:
onFlush and postFlush Entity Listeners

Simply check if an onFlush or postFlush entity listener is subscribed
for the current class in the entity manager during dispatch.

alexgurrola committed:
onFlush and postFlush Entity Listener Tests

Simply run all normal tests for onFlush and postFlush Entity Listeners
based on preFlush Entity Listener checks.

alexgurrola commented:
@mwansinck This Pull Request was held back due to lack of tests, but I have attempted to put the correct bindings for these new Entity Listener events.  Sadly, rebasing caused me to have to clean the branch and rebuild based on the files altered, thus closing and reopening this Pull Request.

alexgurrola commented:
@lcobucci @Ocramius: I attempted to add bindings to the branch for Travis CI, but I'm not sure if I've done something wrong or not.  I don't _really_ understand Travis, so is there some way that I can run some of those checks locally or something?  I'm not really sure how to go forward with the Testing Functionality...  😕

Majkl578 commented:
@alexgurrola Travis is failing because there are test failures caused by your changes. You can see the details when you open a single job (normally there are multiple jobs for different PHP versions and RDBMS) and scroll through the log. I.e. here: https://travis-ci.org/doctrine/doctrine2/jobs/272685091#L595 is failure under PHP 7.1 and SQLite.
You can run tests locally simply by running vendor/bin/phpunit, these will run with SQLite by default. To run against different RDBMS, you would need to copy phpunit.xml.dist to phpunit.xml, uncomment and adjust `db_` variables at bottom. I'd suggest to stick with defaults when not testing DB-specific code (not this case) and leave the rest on Travis.

alexgurrola committed:
Fix Listener Test Function Names

Listener Tests were pointing to incorrect function names.

alexgurrola committed:
Repair Test Enumeration

ostrolucky commented:
continues in https://github.com/doctrine/doctrine2/pull/7310
